R/git.R

```diff
@@ -132,6 +132,102 @@ use_git_config <- function(scope = c("user", "project"), ...) {
   }
 }
 
+#' Choose git protocol and credential defaults
+#'
+#' Decide whether to use the SSH or HTTPS protocol for the current session.
+#' This affects two things:
+#'   * The URL format of newly configured git remotes:
+#'     - `protocol = "ssh"` implies `git@@github.com:<OWNER>/<REPO>.git`
+#'     - `protocol = "https"` implies `https://github.com/<OWNER>/<REPO>.git`
+#'   * The strategy for creating default credentials when none are given.
+#'     Credentials are needed for git operations like `git push` that address a
+#'     remote, typically GitHub. Remember you can always pass explicit
+#'     `credentials` made with [git2r::cred_ssh_key()], [git2r::cred_user_pass],
+#'      and friends.
+#' This function also gives a reminder of how to set a default protocol via an
+#' option: `options(usethis.protocol = "ssh")` or
+#' `options(usethis.protocol = "https")`.
+#'
+#' @section Default credentials:
+#'
+#' The current git protocol is consulted to create default credentials when
+#' they are needed but left unspecified.
+#'
+#' For `protocol = "ssh"`, usethis passes `NULL` credentials straight through to
+#' git2r. This will do the right thing if you have the exact configuration
+#' expected by git2r: your public and private keys are in the default locations,
+#' `~/.ssh/id_rsa.pub` and `~/.ssh/id_rsa`, respectively, and your `ssh-agent`
+#' is configured to manage any associated passphrase, if relevant. Read more
+#' about SSH setup in [Happy Git](http://happygitwithr.com/ssh-keys.html),
+#' especially the [troubleshooting
+#' section](http://happygitwithr.com/ssh-keys.html#ssh-troubleshooting).
+#'
+#' For `protocol = "https"`, usethis attempts to use a GitHub personal access
+#' token (PAT), with preference for an `auth_token` that is passed explicitly.
+#' If that is `NULL`, the environment variables `GITHUB_PAT` and `GITHUB_TOKEN`
+#' are consulted, in that order. If a PAT is found, we make an HTTPS credential
+#' with [git2r::cred_user_pass()]. The PAT is sent as the password and dummy
+#' text is sent as the username (only the PAT is truly required).
+#'
+#' @param protocol "ssh" or "https". Defaults to the option `usethis.protocol`
+#'   and, if unset, to an interactive choice or, in non-interactive sessions,
+#'   "ssh".
+#'
+#' @return "ssh" or "https"
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' use_git_protocol()
+#' use_git_protocol("ssh")
+#' use_git_protocol("https")
+#' }
+use_git_protocol <- function(protocol = NULL) {
```

Comment:
My only concern is that `use` might not be the right verb here. Isn't it more like `discover` or `determine`?

Reply:
You really can call this proactively as `use_git_protocol("https")`, though.

In my mind, in a workshop setting, we might urge everyone to do that right before we try something like `use_github()`. Or I could see using it that way in docs or a non-interactive script.

But IRL, yeah, you would normally just let the menu happen to you upon first need. Or set the option.

I don't feel strongly except that `discover` or `determine` don't reflect the direct usage that is part of the design.

Reply:
Ah ok, in that case it's confusing that we also use it internally to _retrieve_ the current protocol.